setup.py

```diff
@@ -2,21 +2,23 @@
 import os
 from subprocess import check_call
 
-from distutils.core import setup
-from distutils.extension import Extension
+from setuptools import setup
+from setuptools.extension import Extension
 from Cython.Distutils import build_ext
 from Cython.Build import cythonize
-from numpy.distutils.system_info import default_include_dirs, default_lib_dirs
 
-from distutils.sysconfig import get_config_vars
+
+default_include_dirs = []
+default_lib_dirs = []
```

Comment:
@deinst presumably something better needs to go here but I'm not sure what. I think that you said you worked out some code for this...

Reply:
Ive spent a bit of time rummaging about in the numpy.distutils code to see what it does to fill the default_*_dirs variables.  The above works for me, but that is because I keep my flint in a non standard place and need to supply the directory to setup.py.

On the unix/osx side, this is easy, just make a list of the usual suspects, and filter out the ones that don't exist.  On the windows side things are more complicated and I'm working on figuring out  what is needed.

I believe you still need get `config_vars` for the python dirs, but you can get it from `sysconfig` instead of `distutils.sysconfig`

Reply:
> I believe you still need get `config_vars` for the python dirs, but you can get it from `sysconfig` instead of `distutils.sysconfig`

In the version of setup.py on master this is only used to set the `OPT` environment variable:
https://github.com/flintlib/python-flint/blob/2fd4fd6d4a297eed95a3e81082a7df1de605c826/setup.py#L38-L39
What uses that environment variable?

The value I have locally on Linux is:
```python
In [4]: from distutils.sysconfig import get_config_vars

In [5]: get_config_vars('OPT')
Out[5]: ['-DNDEBUG -g -fwrapv -O3 -Wall']
```

Reply:
As I understand it `OPT` supplies default options to gcc.  All setup.py does is removes `-Wstrict-prototypes` from it.  Considering the number of compiler warnings already generated, it seems that filtering out `-Wstrict_prototypes` is probably not particularly helpful.

Reply:
With Python 3.12 I have:
```python
In [2]: from sysconfig import get_config_vars

In [3]: get_config_vars('OPT')
Out[3]: ['-DNDEBUG -g -O3 -Wall']
```
Either way there does not seem to be any `-Wstrict-prototypes`.

Probably that can be removed but I'll leave it there for now while getting the main problems fixed.

I think I might have get the setup.py working for all Python version or OS combinations. Fingers crossed that the Windows build is about to succeed in CI...

The only thing currently missing is `default_include_dirs` and `default_lib_dirs` but I'm not sure why we should need to get those anyway. I would have thought that it was setuptools job to handle the default dirs.

Reply:
If setuptools can get the directories correctly that is great.  If not, I think I have gotten the requisite code from numpy.distutils.system_info to generate the default paths.  I have only a foggy notion of exactly what the windows code is doing as things like `vcpkg` are inovations that postdate my working on code for windows.

Reply:
> I think I have gotten the requisite code from numpy.distutils.system_info to generate the default paths.

I think let's get this PR in for now because it at least means we have CI built wheels for 3.12 and that building a development checkout for 3.12 now works.

We can follow up with another PR for generating the paths. I don't know whether there is actually a reason for doing something other than just trusting setuptools to do it for us though.

Longer term if we can move to meson and meson-python then they should take care of default paths and library discovery etc so in principle it should not be a concern for us.